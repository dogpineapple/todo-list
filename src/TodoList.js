import React, { useState } from 'react';
import NewTodoForm from './NewTodoForm';
import { v4 as uuid } from 'uuid';
import Todo from './Todo';

function TodoList() {
  const [tasks, setTasks] = useState([]);

  const generateTasks = () => {
    return tasks.map(task => (
      <Todo key={task.id}
        id={task.id}
        task={task.task}
        completed={task.completed}
        editTask={editTask}
        deleteTask={deleteTask}
        completeTask={completeTask}
        isEdit={false} />
    ));
  }

  const addTask = (data) => {
    data.id = uuid();
    data.completed = false;
    setTasks(currTasks => [...currTasks, data]);
  }

  const deleteTask = (id) => {
    setTasks(currTasks => currTasks.filter(currTask => currTask.id !== id));
  }

  const editTask = (data, id) => {
    let taskIdx = tasks.findIndex(task => task.id === id);

    setTasks((currTasks) => {
      let editTask = [...currTasks];
      editTask[taskIdx] = { task: data, id, completed: editTask[taskIdx].completed };

      return editTask;
    });
  }

  const completeTask = (id) => {
    let taskIdx = tasks.findIndex(task => task.id === id);

    setTasks((currTasks) => {
      let completeTask = [...currTasks];
      completeTask[taskIdx] = { task: completeTask[taskIdx].task, id, completed: !completeTask[taskIdx].completed };

      return completeTask;
    });
    console.log(tasks);
  }

  return (
    <div>
      <NewTodoForm addTask={addTask} />
      <ul>
        {generateTasks()}
      </ul>
    </div>
  );
}

export default TodoList;