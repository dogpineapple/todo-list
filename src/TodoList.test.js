import React from 'react';
import { render, fireEvent, getByText } from '@testing-library/react';
import TodoList from './TodoList';

it('renders without crashing', () => {
  render(<TodoList />);
});

it('matches snapshot', () => {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it('input changes when value of input changes', () => {
  const { getByLabelText, queryByText } = render(<TodoList />);
  const taskInput = getByLabelText("Create task");
  fireEvent.change(taskInput, { target: { value: "wooo" } });

  expect(taskInput.value).toEqual("wooo");
});

it('submits correctly with task', () => {
  const { getByLabelText, queryByText, getByText } = render(<TodoList />);
  const taskInput = getByLabelText("Create task");
  const submitBtn = getByText("Create");
  fireEvent.change(taskInput, { target: { value: "yes" } });
  fireEvent.click(submitBtn);

  expect(queryByText("yes")).toBeInTheDocument();
});

it('adds multiple tasks', () => {
  const { getByLabelText, queryByText, getByText } = render(<TodoList />);
  const taskInput = getByLabelText("Create task");
  const submitBtn = getByText("Create");
  fireEvent.change(taskInput, { target: { value: "ok" } });
  fireEvent.click(submitBtn);
  fireEvent.change(taskInput, { target: { value: "yes" } });
  fireEvent.click(submitBtn);
  fireEvent.change(taskInput, { target: { value: "yup" } });
  fireEvent.click(submitBtn);

  expect(queryByText("yes")).toBeInTheDocument();
  expect(queryByText("ok")).toBeInTheDocument();
  expect(queryByText("yup")).toBeInTheDocument();
});

it('deletes a task', () => {
  const { getByLabelText, queryByText, getByText } = render(<TodoList />);
  const taskInput = getByLabelText("Create task");
  const submitBtn = getByText("Create");

  fireEvent.change(taskInput, { target: { value: "wheat" } });
  fireEvent.click(submitBtn);

  expect(queryByText("wheat")).toBeInTheDocument();

  const deleteBtn = getByText("X");

  fireEvent.click(deleteBtn);
  expect(queryByText("wheat")).not.toBeInTheDocument();
})