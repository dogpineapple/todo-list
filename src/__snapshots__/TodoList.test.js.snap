// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="task"
      >
        Create task
      </label>
      <input
        id="task"
        name="task"
        value=""
      />
      <button>
        Create
      </button>
    </form>
    <ul />
  </div>
</DocumentFragment>
`;
